### PR TITLE
label maker v0 implementation

### PR DESCRIPTION
**Pull request by bnewbold**

### CODE REVIEW

File: cmd/labelmaker/main.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		cs, err := carstore.NewCarStore(carstdb, "data/labelmaker/carstore")
+		if err != nil {
+			return err
+		}
+
+		//labelmakerhost := cctx.String("labelmakerhost")
+		repoDid := "did:plc:FAKE"
+		repoHandle := "labelmaker.test"
+		//bgsUrl := "ws://localhost:2470/events"
+		//bgsUrl := "ws://[::1]:4989/events"
```
Comment (devinivy) [inline cmd/labelmaker/main.go:119]: The way things are looking right now, I believe these will be xrpc endpoints spec'ed with lexicons (as will the output of the label-maker).  So the endpoints may look something like `ws://localhost:2470/xrpc/com.atproto.sync.stream`, `ws://localhost:2470/xrpc/com.atproto.label.stream`.

File: events/label_events.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+// this is here, instead of under 'labeling' package, to avoid an import loop
+type Label struct {
+	LexiconTypeID string  `json:"$type" cborgen:"$type,const=app.bsky.label.label"`
+	SourceDid     string  `json:"src" cborgen:"src"`
+	SubjectUri    string  `json:"uri" cborgen:"uri"`
+	SubjectCid    *string `json:"cid,omitempty" cborgen:"cid"`
+	Value         string  `json:"val" cborgen:"val"`
+	Timestamp     string  `json:"ts" cborgen:"ts"` // TODO: actual timestamp?
```
Comment (devinivy) [inline events/label_events.go:84]: Re: the TODO, hard question!  I think ideally we use lex's datetime type for this in order to cut with the grain of lexicon, and if it needs to be more compact for some reason then we address that through lex/encodings in the future.

Comment (bnewbold) [reply events/label_events.go:84]: Yeah, the plan here is to use codegen'd type from `com.atproto.label.label`, and inherit the correct type on this field from there.

Cross-referencing to https://github.com/bluesky-social/atproto/pull/525, where there is discussion about `labelUri` field.

File: cmd/labelmaker/main.go (hunk 1 above)
Comment (bnewbold) [reply cmd/labelmaker/main.go:112]: yup, sounds good!
Also these commented things were cruft from not understanding what the format of the "URI" was supposed to be, about to overwrite this.

File: labeling/ws_endpoints.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			return err
+		}
+
+		if err := header.MarshalCBOR(wc); err != nil {
+			return fmt.Errorf("failed to write header: %w", err)
+		}
+
+		if err := evt.MarshalCBOR(wc); err != nil {
+			return fmt.Errorf("failed to write event: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply labeling/ws_endpoints.go:48]: I'm not sure what you mean by "interleaved". Like they get off-by-one and a header gets mis-interpreted as a body or something?
This code is basically copypasta from why's repoevent stream code, and will likely get updated to match any changes there (or abstracted away to reduce duplication) soon.

[the remark this answers is not in the document]
Comment (devinivy) [reply labeling/ws_endpoints.go:48]: Yeah, let me try to rephrase that!  Say `evts` produces two events, the first with `header1` and `body1`, and the second with `header2` and `body2`.  Then we want those to go over the wire `header1,body1,header2,body2`.  I noticed that the writes of header and body are separate here, so I am wondering if there's any way they could end-up going over the wire as e.g. `header1,header2,body1,body2`.  My guess is that iterating over `evts` here ensures one `evt` is processed at a time, but just wasn't sure!

[the remark this answers is not in the document]
Comment (bnewbold) [reply labeling/ws_endpoints.go:48]: Ah, ok, I think I gotcha.

There are two layers of abstraction preventing something like a race-condition on publishing:

1) this code, running in `EventsLabelsWebsocket()`, runs with only single concurrency per websocket connection. there might be multiple labelers running concurrently and pushing events to the `events.LabelEventManager`, and that manager might be fanning out those events over separate channels to multiple websocket handler routines, but for any single websocket connection there is only one handler running. So there is no race on writes to the `conn`.

2) entirely separate from that, websocket messages are framed, and binary label messages should easily be small enough to not get fragmented. the "gorilla/websocket" API starts a frame with `conn.NextWriter()`, data and written in to buffer, and ends at `wc.Close()`, when the buffer is sent as a single frame. so even if there were concurrent writers to the single websocket connection, the header/body pairings should never end up interleaved. I haven't dug in to the details to verify, but based on the websocket spec, I would be surprised if "gorilla/websocket" ever resulted in interleaving, even with concurrent fragmented frames (single large binary buffer getting split into multiple sequential frames).

[the remark this answers is not in the document]
Comment (devinivy) [reply labeling/ws_endpoints.go:48]: Right on, thanks for laying that out!

**Comment by bnewbold** [comment]
Have an updated/rebased version of this ready for actual review: https://github.com/bluesky-social/indigo/pull/60

**Pull request closed without merging**